### PR TITLE
Fix up-to-date check race condition

### PR DESCRIPTION
**Pull request by drewnoakes**
(Includes PR #4022, so just review 73438f95b3b52c101f430477bc5c851817298d7e.)

Issue #4014 describes a race condition. This PR fixes that race by introducing an additional check. If an input has been modified since the last time a check was requested, then the project is considered _not_ up to date.

### CODE REVIEW

**Comment by drewnoakes** [comment]
@dotnet-bot retest windows_debug please

**Comment by drewnoakes** [comment]
@dotnet-bot test this please

**Comment by drewnoakes** [comment]
@Pilchie using UTC throughout seems to have worked well. Does it look ok to you?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/IFileSystemMock.cs [hunk 1]
```diff
@@ -205,28 +203,32 @@ public void WriteAllText(string path, string content)
...
             }
             else
             {
-                _files[path] = new FileData() { FileContents = content, FileEncoding = encoding };
-                SetLastWriteTime(_files[path]);
+                Files[path] = new FileData
+                {
+                    FileContents = content,
+                    FileEncoding = encoding,
+                    LastWriteTimeUtc = DateTime.UtcNow
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/IFileSystemMock.cs]: The call to `SetLastWriteTime` wasn't important?

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/IFileSystemMock.cs]: On a new instance it is equivalent.

Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/IFileSystemMock.cs]: Great!

**Pull request merged**